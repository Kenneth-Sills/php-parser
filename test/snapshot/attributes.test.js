const parser = require("../main");

describe("Parse Attributes", () => {
  it("can parse class attributes", () => {
    expect(
      parser.parseEval(`
    #[Deprecated]
    #[replace("use NewClass")]
    class DepClass {}
    `)
    ).toMatchSnapshot();
  });
  it("can parse method attributes", () => {
    expect(
      parser.parseEval(`
    class Test {
      #[Pure]
      function m() {}
    }
    `)
    ).toMatchSnapshot();
  });
  it("can parse param attributes", () => {
    expect(
      parser.parseEval(`
    function f(
    #[Unsigned]
     int $n) {}
    `)
    ).toMatchSnapshot();
  });
  it("can parse params with comments", () => {
    expect(
      parser.parseEval(
        `// Line 1
    #[ Pure ( )/* Pure */ ]
    // Line 3
    function a(#[ Unsigned ] $a) {}
    `,
        {
          parser: {
            extractDoc: true,
          },
        }
      )
    ).toMatchSnapshot();
  });
  it("can parse parms with array values", () => {
    expect(
      parser.parseEval(`
    #[List(["a"=>1, 'b' => Test::class])]
    function a() {}
    `)
    ).toMatchSnapshot();
  });
  it("can parse params with end characters", () => {
    expect(
      parser.parseEval(`
    #[End(["])}>"])]
    class End {}
    `)
    ).toMatchSnapshot();
  });
  it("can parse multi-line attributes", () => {
    expect(
      parser.parseEval(`
    #[
      One(),
      Two(),
      Three()
    ]
    #[Four]
    class Multi {}
    `)
    ).toMatchSnapshot();
  });
});
