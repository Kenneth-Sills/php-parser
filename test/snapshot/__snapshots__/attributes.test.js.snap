// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse Attributes can parse class attributes 1`] = `
Program {
  "children": Array [
    Class {
      "attrGroups": Array [
        AttrGroup {
          "attrs": Array [
            Attribute {
              "args": Array [],
              "kind": "attribute",
              "name": "Deprecated",
            },
          ],
          "kind": "attrgroup",
        },
        AttrGroup {
          "attrs": Array [
            Attribute {
              "args": Array [
                String {
                  "isDoubleQuote": true,
                  "kind": "string",
                  "raw": "\\"use NewClass\\"",
                  "unicode": false,
                  "value": "use NewClass",
                },
              ],
              "kind": "attribute",
              "name": "replace",
            },
          ],
          "kind": "attrgroup",
        },
      ],
      "body": Array [],
      "extends": null,
      "implements": null,
      "isAbstract": false,
      "isAnonymous": false,
      "isFinal": false,
      "kind": "class",
      "name": Identifier {
        "kind": "identifier",
        "name": "DepClass",
      },
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;

exports[`Parse Attributes can parse method attributes 1`] = `
Program {
  "children": Array [
    Class {
      "attrGroups": Array [],
      "body": Array [
        Method {
          "arguments": Array [],
          "attrGroups": Array [
            AttrGroup {
              "attrs": Array [
                Attribute {
                  "args": Array [],
                  "kind": "attribute",
                  "name": "Pure",
                },
              ],
              "kind": "attrgroup",
            },
          ],
          "body": Block {
            "children": Array [],
            "kind": "block",
          },
          "byref": false,
          "isAbstract": false,
          "isFinal": false,
          "isStatic": false,
          "kind": "method",
          "name": Identifier {
            "kind": "identifier",
            "name": "m",
          },
          "nullable": false,
          "type": null,
          "visibility": "",
        },
      ],
      "extends": null,
      "implements": null,
      "isAbstract": false,
      "isAnonymous": false,
      "isFinal": false,
      "kind": "class",
      "name": Identifier {
        "kind": "identifier",
        "name": "Test",
      },
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;

exports[`Parse Attributes can parse multi-line attributes 1`] = `
Program {
  "children": Array [
    Class {
      "attrGroups": Array [
        AttrGroup {
          "attrs": Array [
            Attribute {
              "args": Array [],
              "kind": "attribute",
              "name": "One",
            },
            Attribute {
              "args": Array [],
              "kind": "attribute",
              "name": "Two",
            },
            Attribute {
              "args": Array [],
              "kind": "attribute",
              "name": "Three",
            },
          ],
          "kind": "attrgroup",
        },
        AttrGroup {
          "attrs": Array [
            Attribute {
              "args": Array [],
              "kind": "attribute",
              "name": "Four",
            },
          ],
          "kind": "attrgroup",
        },
      ],
      "body": Array [],
      "extends": null,
      "implements": null,
      "isAbstract": false,
      "isAnonymous": false,
      "isFinal": false,
      "kind": "class",
      "name": Identifier {
        "kind": "identifier",
        "name": "Multi",
      },
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;

exports[`Parse Attributes can parse param attributes 1`] = `
Program {
  "children": Array [
    _Function {
      "arguments": Array [
        Parameter {
          "attrGroups": Array [
            AttrGroup {
              "attrs": Array [
                Attribute {
                  "args": Array [],
                  "kind": "attribute",
                  "name": "Unsigned",
                },
              ],
              "kind": "attrgroup",
            },
          ],
          "byref": false,
          "flags": 0,
          "kind": "parameter",
          "name": Identifier {
            "kind": "identifier",
            "name": "n",
          },
          "nullable": false,
          "type": TypeReference {
            "kind": "typereference",
            "name": "int",
            "raw": "int",
          },
          "value": null,
          "variadic": false,
        },
      ],
      "attrGroups": Array [],
      "body": Block {
        "children": Array [],
        "kind": "block",
      },
      "byref": false,
      "kind": "function",
      "name": Identifier {
        "kind": "identifier",
        "name": "f",
      },
      "nullable": false,
      "type": null,
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;

exports[`Parse Attributes can parse params with comments 1`] = `
Program {
  "children": Array [
    _Function {
      "arguments": Array [
        Parameter {
          "attrGroups": Array [
            AttrGroup {
              "attrs": Array [
                Attribute {
                  "args": Array [],
                  "kind": "attribute",
                  "name": "Unsigned",
                },
              ],
              "kind": "attrgroup",
            },
          ],
          "byref": false,
          "flags": 0,
          "kind": "parameter",
          "name": Identifier {
            "kind": "identifier",
            "name": "a",
          },
          "nullable": false,
          "type": null,
          "value": null,
          "variadic": false,
        },
      ],
      "attrGroups": Array [
        AttrGroup {
          "attrs": Array [
            CommentLine {
              "kind": "commentline",
              "offset": 0,
              "value": "// Line 1
",
            },
            Attribute {
              "args": Array [],
              "kind": "attribute",
              "leadingComments": Array [
                CommentBlock {
                  "kind": "commentblock",
                  "offset": 26,
                  "value": "/* Pure */",
                },
              ],
              "name": "Pure",
            },
          ],
          "kind": "attrgroup",
        },
      ],
      "body": Block {
        "children": Array [],
        "kind": "block",
      },
      "byref": false,
      "kind": "function",
      "leadingComments": Array [
        CommentLine {
          "kind": "commentline",
          "offset": 43,
          "value": "// Line 3
",
        },
      ],
      "name": Identifier {
        "kind": "identifier",
        "name": "a",
      },
      "nullable": false,
      "type": null,
    },
  ],
  "comments": Array [
    CommentLine {
      "kind": "commentline",
      "offset": 0,
      "value": "// Line 1
",
    },
    CommentBlock {
      "kind": "commentblock",
      "offset": 26,
      "value": "/* Pure */",
    },
    CommentLine {
      "kind": "commentline",
      "offset": 43,
      "value": "// Line 3
",
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;

exports[`Parse Attributes can parse params with end characters 1`] = `
Program {
  "children": Array [
    Class {
      "attrGroups": Array [
        AttrGroup {
          "attrs": Array [
            Attribute {
              "args": Array [
                Array {
                  "items": Array [
                    Entry {
                      "byRef": false,
                      "key": null,
                      "kind": "entry",
                      "unpack": false,
                      "value": String {
                        "isDoubleQuote": true,
                        "kind": "string",
                        "raw": "\\"])}>\\"",
                        "unicode": false,
                        "value": "])}>",
                      },
                    },
                  ],
                  "kind": "array",
                  "shortForm": true,
                },
              ],
              "kind": "attribute",
              "name": "End",
            },
          ],
          "kind": "attrgroup",
        },
      ],
      "body": Array [],
      "extends": null,
      "implements": null,
      "isAbstract": false,
      "isAnonymous": false,
      "isFinal": false,
      "kind": "class",
      "name": Identifier {
        "kind": "identifier",
        "name": "End",
      },
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;

exports[`Parse Attributes can parse parms with array values 1`] = `
Program {
  "children": Array [
    _Function {
      "arguments": Array [],
      "attrGroups": Array [
        AttrGroup {
          "attrs": Array [
            Attribute {
              "args": Array [
                Array {
                  "items": Array [
                    Entry {
                      "byRef": false,
                      "key": String {
                        "isDoubleQuote": true,
                        "kind": "string",
                        "raw": "\\"a\\"",
                        "unicode": false,
                        "value": "a",
                      },
                      "kind": "entry",
                      "unpack": false,
                      "value": Number {
                        "kind": "number",
                        "value": "1",
                      },
                    },
                    Entry {
                      "byRef": false,
                      "key": String {
                        "isDoubleQuote": false,
                        "kind": "string",
                        "raw": "'b'",
                        "unicode": false,
                        "value": "b",
                      },
                      "kind": "entry",
                      "unpack": false,
                      "value": StaticLookup {
                        "kind": "staticlookup",
                        "offset": Identifier {
                          "kind": "identifier",
                          "name": "class",
                        },
                        "what": Name {
                          "kind": "name",
                          "name": "Test",
                          "resolution": "uqn",
                        },
                      },
                    },
                  ],
                  "kind": "array",
                  "shortForm": true,
                },
              ],
              "kind": "attribute",
              "name": "List",
            },
          ],
          "kind": "attrgroup",
        },
      ],
      "body": Block {
        "children": Array [],
        "kind": "block",
      },
      "byref": false,
      "kind": "function",
      "name": Identifier {
        "kind": "identifier",
        "name": "a",
      },
      "nullable": false,
      "type": null,
    },
  ],
  "errors": Array [],
  "kind": "program",
}
`;
